UniversalAlgebra/Structures/Graphs.lagda

```diff
@@ -0,0 +1,199 @@
+
+open signature
+open structure
+open _⊎_
+
+ℓ₁ : Level
+ℓ₁ = lsuc ℓ₀
+
+Gr-sig : signature → signature → signature
+Gr-sig 𝐹 𝑅 = record { symbol = symbol 𝑅 ⊎ symbol 𝐹 ; arity = arty }
+ where
+ arty : symbol 𝑅 ⊎ symbol 𝐹 → Type ℓ₀
+ arty (inl 𝑟) = (arity 𝑅) 𝑟
+ arty (inr 𝑓) = (arity 𝐹) 𝑓 ⊎ 𝟙
+
+
+module _ {𝐹 𝑅 : signature} where
+
+ Gr : structure 𝐹 {ℓ₀} 𝑅 {ℓ₀} → structure Sig∅ {ℓ₀} (Gr-sig 𝐹 𝑅) {ℓ₀}
+ Gr 𝑨 = record { carrier = carrier 𝑨 ; op = λ () ; rel = split }
+  where
+  split : (s : symbol 𝑅 ⊎ symbol 𝐹) → Rel (carrier 𝑨) (arity (Gr-sig 𝐹 𝑅) s) {ℓ₀}
+  split (inl 𝑟) arg = rel 𝑨 𝑟 arg
+  split (inr 𝑓) args = op 𝑨 𝑓 (args ∘ inl) ≡ args (inr 𝟙.𝟎)
+
+
+open ≡-Reasoning
+
+module _ {𝐹 𝑅 : signature}
+         {𝑨 𝑩 : structure 𝐹 {ℓ₀} 𝑅 {ℓ₀}} where
+
+ hom→Grhom : hom 𝑨 𝑩 → hom (Gr 𝑨) (Gr 𝑩)
+ hom→Grhom (h , hhom) = h , (i , ii)
+  where
+  i : is-hom-rel (Gr 𝑨) (Gr 𝑩) h
+  i (inl 𝑟) a x = ∣ hhom ∣ 𝑟 a x
+  i (inr 𝑓) a x = goal
+   where
+   homop : h (op 𝑨 𝑓 (a ∘ inl)) ≡ op 𝑩 𝑓 (h ∘ (a ∘ inl))
+   homop = (snd hhom) 𝑓 (a ∘ inl)
+
+   goal : op 𝑩 𝑓 (h ∘ (a ∘ inl)) ≡ h (a (inr 𝟙.𝟎))
+   goal = op 𝑩 𝑓 (h ∘ (a ∘ inl)) ≡⟨ sym homop ⟩
+          h (op 𝑨 𝑓 (a ∘ inl))   ≡⟨ cong h x ⟩
+          h (a (inr 𝟙.𝟎))         ∎
+
+  ii : is-hom-op (Gr 𝑨) (Gr 𝑩) h
+  ii = λ ()
+
+
+ Grhom→hom : hom (Gr 𝑨) (Gr 𝑩) → hom 𝑨 𝑩
+ Grhom→hom (h , hhom) = h , (i , ii)
+  where
+  i : is-hom-rel 𝑨 𝑩 h
+  i R a x = fst hhom (inl R) a x
+  ii : is-hom-op 𝑨 𝑩 h
+  ii f a = goal
+   where
+   split : arity 𝐹 f ⊎ 𝟙 → carrier 𝑨
+   split (inl x) = a x
+   split (inr y) = op 𝑨 f a
+   goal : h (op 𝑨 f a) ≡ op 𝑩 f (λ x → h (a x))
+   goal = sym (fst hhom (inr f) split refl)
+
+
+
+{- Lemma III.1. Let S be a signature and A be an S-structure.
+Let Σ be a finite set of identities such that A ⊧ Σ. For every
+instance X of CSP(A), one can compute in polynomial time an
+instance Y of CSP(A) such that Y ⊧ Σ and | Hom(X , A)| = |Hom(Y , A)|. -}
+
+
+module _ {𝐹 : signature}{χ : Level} where
+
+ data Term (X : Type χ ) : Type χ  where
+  ℊ : X → Term X    -- (ℊ for "generator")
+  node : (f : symbol 𝐹)(𝑡 : (arity 𝐹) f → Term X) → Term X
+
+ open Term public
+
+\end{code}
+
+When we interpret a term in a structure we call the resulting function a *term operation*.  Given a term `p` and a structure `𝑨`, we denote by `𝑨 ⟦ p ⟧` the *interpretation* of `p` in `𝑨`.  This is defined inductively as follows.
+
+1. If `p` is a variable symbol `x : X` and if `a : X → ∣ 𝑨 ∣` is a tuple of elements of `∣ 𝑨 ∣`, then `𝑨 ⟦ p ⟧ a := a x`.
+
+2. If `p = 𝑓 𝑡`, where `𝑓 : ∣ 𝑆 ∣` is an operation symbol, if `𝑡 : ∥ 𝑆 ∥ 𝑓 → 𝑻 X` is a tuple of terms, and if `a : X → ∣ 𝑨 ∣` is a tuple from `𝑨`, then we define `𝑨 ⟦ p ⟧ a = 𝑨 ⟦ 𝑓 𝑡 ⟧ a := (𝑓 ̂ 𝑨) (λ i → 𝑨 ⟦ 𝑡 i ⟧ a)`.
+
+Thus the interpretation of a term is defined by induction on the structure of the term, and the definition is formally implemented as follows.
+
+\begin{code}
+
+module _ {𝐹 𝑅 : signature}{χ : Level}{X : Type χ} where
```

Comment:
This whole module is nice - but it looks identical to something I've seen elsewhere in this same code. Why the duplication?

Reply:
Did I submit duplicate code by accident?  I don't think we do conversion of operations to relations (graphs) anywhere else... but I might have overlooked something.

Reply:
Ah, now I see what you mean.  Yes, I duplicated the code implementing entailment/models because I'm using structures in this file instead of Algebras or algebras or Algebroids or SetoidAlgebras.

Maybe we should decide which of these algebraic/relational structure type(s) is (are) the best and delete the others, along with the duplicate code for them, or alternatively try to make the other types (for entailment/models, homs, congruences, etc) more abstract/polymorphic so we can use them with different representations of structures.  What do you think?

Reply:
I'm a big fan of first trying polymorphic. If that fails, then making a choice might be the best route.